cassiopeia/solver/PercolationSolver.py

```diff
@@ -0,0 +1,350 @@
+from collections import defaultdict
+import itertools
+import networkx as nx
+import numpy as np
+import pandas as pd
+from typing import Callable, Dict, List, Optional, Tuple, Union
+
+from cassiopeia.data import CassiopeiaTree
+from cassiopeia.data import utilities as data_utilities
+from cassiopeia.solver import CassiopeiaSolver
+from cassiopeia.solver import NeighborJoiningSolver
+from cassiopeia.solver import dissimilarity_functions
+from cassiopeia.solver import graph_utilities
+from cassiopeia.solver import solver_utilities
+
+
+class PercolationSolver(CassiopeiaSolver.CassiopeiaSolver):
+    """
+    TODO: Experiment to find the best default similarity function
+    The PercolationSolver implements a top-down algorithm that recursively
+    partitions the sample set based on similarity in the observed mutations.
+    It is an implicit version of Aho's algorithm for tree discovery (1981).
+    At each recursive step, the similarities of each sample pair are embedded
+    in a graph. The graph is then percolated by removing the minimum edges
+    until multiple connected components are produced. The algorithm enforces
+    binary partitions if there are more than two connected components using a
+    neighbor-joining procedure.
+
+    Args:
+        joining_solver: The CassiopeiaSolver that is used to cluster groups of
+            samples in the case that the percolation procedure generates more
+            than two groups of samples in the partition
+        prior_transformation: A function defining a transformation on the priors
+            in forming weights to scale the contribution of each mutation in
+            the similarity graph
+        similarity_function: A function that calculates a similarity score
+            between two given samples and their observed mutations. The default
+            is "hamming_distance_without_missing"
+        threshold: A minimum similarity threshold to include an edge in the
+            similarity graph
```

Comment:
Can you give some intuition here, or somewhere else, what the similarity threshold is supposed to look like? (Like what does 1 mean? Is supposed to be between 0 and 1? etc.) What's a reasonable threshold? Is a good default really 0?